# EDA - Estatistica Descritiva

<a href="../../README.md" title="Voltar para a página principal">
🏠 Voltar para Home
</a>

## Visão Geral

A Estatística Descritiva compreende o conjunto de técnicas voltadas à exploração inicial de um conjunto de dados, com foco na compreensão da distribuição, comportamento, padrões centrais, variabilidade e possíveis anomalias.  
Nesta etapa, o objetivo principal é obter uma leitura clara da estrutura do dataset, identificar problemas de integridade e avaliar características fundamentais que influenciarão diretamente na modelagem preditiva.

As análises contemplam os seguintes pilares:
- identificação de medidas centrais (média, mediana, moda)
- avaliação de dispersão (variância, desvio padrão e coeficiente de variação)
- estudo da forma da distribuição (assimetria e curtose)
- investigação de outliers estatísticos e lógicos
- análise da frequência de variáveis categóricas
- avaliação de correlações entre variáveis numéricas
- análise da integridade e consistência estrutural do dataset

Estas informações formam o alicerce para as próximas etapas do CRISP-DM, especialmente Preparação dos Dados e Modelagem.

## Objetivo

O objetivo desta etapa é aplicar técnicas de **Análise Estatística Descritiva** utilizando Python e as bibliotecas **NumPy** e **Pandas**, com apoio de tabelas e visualizações gráficas.
A análise descritiva tem como finalidade compreender a estrutura, qualidade e comportamento das variáveis do dataset, servindo como base para decisões de pré-processamento, feature engineering e modelagem preditiva.

A partir desta etapa, buscamos responder às seguintes questões fundamentais sobre o conjunto de dados:

1. **Existem valores ausentes (missing values)?**
   Quantos são? Em quais colunas se concentram? Representam ausência legítima ou falha de coleta?

2. **Existem registros duplicados?**
   Devem ser removidos? Qual a regra de unicidade aplicável ao dataset?

3. **Cada variável está corretamente tipada?**
   Datas estão reconhecidas como `datetime`?
   Variáveis numéricas foram importadas como texto?
   Campos categóricos estão mapeados como `object`?

4. **Existem valores inválidos ou logicamente impossíveis?**
   Exemplos: valores negativos em atributos que não deveriam assumir tais valores, datas inconsistentes ou fora de ordem, quantidades incompatíveis com o processo operacional.

5. **Qual é a tendência central das variáveis numéricas?**
   Avaliação de média, mediana e moda, identificando o comportamento central e possíveis distorções.

6. **Qual é o grau de dispersão dos dados?**
   Análise de variância, desvio padrão e coeficiente de variação para verificar estabilidade ou heterogeneidade das medidas.

7. **As variáveis apresentam simetria ou assimetria?**
   Cálculo de *skewness* e *kurtosis* para identificar distribuições alongadas, achatadas ou com caudas extremas.

8. **Como se comportam as distribuições de cada variável?**
   Por meio de histogramas, densidades, boxplots e outras representações visuais.

9. **Existem outliers extremos ou anomalias estatísticas?**
   Identificação via IQR, Z-score ou inspeção gráfica, avaliando o impacto no processo e na futura modelagem.

# Importações

In [387]:
# Manipulação de arquivos e diretórios
import sys
import os

# Dowload Kagglehub
# import kagglehub

# Estatisticas e DataFrames
import pandas as pd
import numpy as np

# Gráficos
import matplotlib.pyplot as plt
import seaborn as sns

# Trabalho com html
from IPython.display import display, HTML

# 1. Carregamento e Visualização Preliminar

***Descrição:*** Utilizando o dataset de acompanhamento operacional dos pedidos tratados.
> **Arquivo e:** database/processed/acompanhamento_operacional_clean.csv

In [388]:
file_path = "../database/processed/acompanhamento_operacional_clean.csv"

df_main = pd.read_csv(file_path)

df_main.head(5)

sigla_cliente       ss tipo_veiculo  qtde_itens  volume    peso    m3  uf  \
0           NTL  3560316  TRUCK 70 M3           1       1  144.30  2.10  MS   
1           NTL  3561720  TRUCK 70 M3           1       1  130.30  2.33  MS   
2           NTL  3562246  TRUCK 70 M3           1       1   97.00  2.22  SC   
3           MMM  3565015  TRUCK 75 M3           7      43  182.79  1.08  CE   
4           MMM  3565022  TRUCK 70 M3           4       9   25.50  0.07  CE   

   fl_base                   representante  ... horas_planejamento  \
0        0                    NÃO DEFINIDO  ...                3.0   
1        0                    NÃO DEFINIDO  ...                3.0   
2        0       SC CARGO TRANSPORTES LTDA  ...                3.0   
3        0  VELOMAX BRASIL TRANSPORTES LTD  ...                2.0   
4        0                 MFM TRANSPORTES  ...                2.0   

  horas_divisao_ocam horas_coleta  horas_conferencia  horas_emissao  \
0               20.0          6.0                0.0            0.0   
1               20.0          7.0                0.0            0.0   
2              187.0          2.0                0.0            0.0   
3               46.0          4.0                1.0            0.0   
4               46.0          6.0                1.0            0.0   

   horas_analise_producao horas_minuta  horas_exped_minuta  \
0                    22.0          0.0                16.0   
1                    22.0          0.0                16.0   
2                     0.0          0.0                16.0   
3                    11.0         99.0                 8.0   
4                    12.0        172.0                16.0   

  hora_analise_transporte  analise_transporte  
0                   161.0       Fora do Prazo  
1                   165.0       Fora do Prazo  
2                    40.0       Fora do Prazo  
3                    93.0       Fora do Prazo  
4                   285.0       Fora do Prazo  

[5 rows x 32 columns]

# 2. Tendencias Centrais - Medidas de Centralidade (Média, Mediana, Moda)

***Descrição:*** Entender a natureza e a distribuição das variáveis.  
As tendências centrais, ou medidas de centralidade, são valores únicos que representam o centro de um conjunto de dados, sendo a média, a mediana e a moda as mais comuns. Elas servem para resumir dados de forma mais simples, como a idade de um grupo de 100 pessoas, salario de um grupo de funcionarios, etc. A escolha de qual medida usar depende da natureza dos dados, como a presença de valores extremos que podem influenciar a média. 

**Medidas e Visualizações**
- Média
- mediana
- moda

> - **Se mean e 50% (mediana)** estão próximos → boa simetria.
> - **Se min e max** estão muito distantes → alto desvio padrão (talvez outliers).
> - **Se std é da mesma ordem** de grandeza da média → alta variabilidade.

## Visão Geral das Medidas de Centralidade

- Avaliar e classificar as medidas de centralidade das 17 features quantitativas
- Avaliar e classificar o tamanho da cauda e classificar cada featere como: Simétrica, assimétrica esquerda, assimétria a direita
- Avaliar o desvio padrão vs medias, verificar as proximidades de cada feature
- Verificar skewness e kurtosis de cada feature

In [389]:
df_num = df_main.select_dtypes(include=['int64','float64'])
cols_remover = ['ss','fl_base','fl_atraso_zt','fl_atraso_cli','fl_sem_minuta',]
df_num = df_num.drop(columns=cols_remover)
df_num.describe().T.sort_values(by='std', ascending=False)

count        mean          std  min    25%  \
peso_cubado_rodoviario    488398.0  342.633352  3768.429286  0.0   2.00   
hora_analise_transporte   488398.0  152.443169   766.479116  0.0  22.00   
horas_minuta              488398.0   91.374789   309.359113  0.0  16.00   
horas_emissao             488398.0   27.527525   308.377685  0.0   0.00   
peso                      488398.0   44.923519   283.198508  0.0   1.50   
volume                    488398.0   10.194352   228.325868  1.0   1.00   
horas_analise_producao    488398.0   43.872112    71.470168  0.0   6.00   
horas_divisao_ocam        488398.0   42.608809    64.533771  0.0   2.00   
horas_pre_conferencia     488398.0   16.057553    61.856582  0.0   0.00   
horas_exped_minuta        488398.0   19.335800    36.355978  0.0   8.00   
horas_coleta              488398.0    9.247427    31.685940  0.0   0.00   
horas_conferencia         488398.0    7.715695    30.907774  0.0   0.00   
horas_planejamento        488398.0    4.638960    19.807807  0.0   0.00   
qtde_itens                488398.0    5.152421     7.952216  1.0   1.00   
qtde_ocams                488398.0    0.973114     3.459025  0.0   0.00   
horas_distribuicao_cotas  488398.0    0.553514     3.433501  0.0   0.00   
m3                        488398.0    0.457469     2.348216  0.0   0.01   

                            50%     75%         max  
peso_cubado_rodoviario    10.00   75.00  1511070.00  
hora_analise_transporte   95.00  217.00   262968.00  
horas_minuta              45.00   96.00    13135.00  
horas_emissao              1.00    2.00    13223.00  
peso                       2.20   17.73    90390.00  
volume                     1.00    3.00    86363.00  
horas_analise_producao    22.00   55.00     8790.00  
horas_divisao_ocam        17.00   62.00     1488.00  
horas_pre_conferencia      1.00    4.00     4078.00  
horas_exped_minuta        10.00   16.00     4954.00  
horas_coleta               1.00    4.00     4269.00  
horas_conferencia          0.00    1.00     1219.00  
horas_planejamento         1.00    4.00     1014.00  
qtde_itens                 2.00    6.00      180.00  
qtde_ocams                 0.00    0.00      105.00  
horas_distribuicao_cotas   0.00    0.00      198.00  
m3                         0.02    0.13      276.48

> 💡 **Nota Técnica:** 
> As estimativas das medidas de centralidade foram aplicadas na base de dados consolidada, ou seja, com a granularidade mais alta, para observar o comportamento geral dos dados.

In [ ]:
# Calcula média e mediana de todas as colunas
df_stats = df_num.agg(['mean', 'median']).T

# Calcula assimetria relativa (%)
df_stats['assimetria_relativa_%'] = (
    (df_stats['mean'] - df_stats['median']).abs() /
    df_stats['median'].replace(0, 1e-6)
) * 100

# Classifica
def classificar_assimetria(x):
    if pd.isna(x): return 'indefinido'
    elif x < 10: return 'Normal'
    elif x < 30: return 'Leve Assimetria'
    elif x < 50: return 'Assimetria Significativa'
    else: return 'Forte Assimetria'

df_stats['classe_assimetria'] = df_stats['assimetria_relativa_%'].apply(classificar_assimetria)

# Ordena pelas mais distorcidas
df_stats.sort_values('assimetria_relativa_%', ascending=False)

mean  median  assimetria_relativa_%  \
horas_conferencia           7.715695    0.00           7.715695e+08   
qtde_ocams                  0.973114    0.00           9.731141e+07   
horas_distribuicao_cotas    0.553514    0.00           5.535137e+07   
peso_cubado_rodoviario    342.633352   10.00           3.326334e+03   
horas_emissao              27.527525    1.00           2.652752e+03   
m3                          0.457469    0.02           2.187343e+03   
peso                       44.923519    2.20           1.941978e+03   
horas_pre_conferencia      16.057553    1.00           1.505755e+03   
volume                     10.194352    1.00           9.194352e+02   
horas_coleta                9.247427    1.00           8.247427e+02   
horas_planejamento          4.638960    1.00           3.638960e+02   
qtde_itens                  5.152421    2.00           1.576210e+02   
horas_divisao_ocam         42.608809   17.00           1.506401e+02   
horas_minuta               91.374789   45.00           1.030551e+02   
horas_analise_producao     43.872112   22.00           9.941869e+01   
horas_exped_minuta         19.335800   10.00           9.335800e+01   
hora_analise_transporte   152.443169   95.00           6.046649e+01   

                         classe_assimetria  
horas_conferencia         Forte Assimetria  
qtde_ocams                Forte Assimetria  
horas_distribuicao_cotas  Forte Assimetria  
peso_cubado_rodoviario    Forte Assimetria  
horas_emissao             Forte Assimetria  
m3                        Forte Assimetria  
peso                      Forte Assimetria  
horas_pre_conferencia     Forte Assimetria  
volume                    Forte Assimetria  
horas_coleta              Forte Assimetria  
horas_planejamento        Forte Assimetria  
qtde_itens                Forte Assimetria  
horas_divisao_ocam        Forte Assimetria  
horas_minuta              Forte Assimetria  
horas_analise_producao    Forte Assimetria  
horas_exped_minuta        Forte Assimetria  
hora_analise_transporte   Forte Assimetria

In [ ]:
df_frec_stats = (
    df_stats['classe_assimetria']
    .value_counts()
    .rename_axis('classe_assimetria')
    .reset_index(name='volume')
)

df_frec_stats

classe_assimetria  volume
0  Forte Assimetria      17

**Nota Ténica:** A análise estrutural das variáveis quantitativas evidencia um padrão estatístico marcante: **todas apresentam forte assimetria à direita**, caracterizada pela presença de **caudas longas** que se estendem para valores muito superiores ao comportamento central da distribuição.

Esse diagnóstico emerge prontamente quando comparamos as medidas de centralidade, especialmente **média** e **mediana**. Observou-se que, para praticamente todas as features operacionais, a **mediana** encontra-se concentrada entre **0 e 10 horas**, enquanto as **médias** assumem valores muito mais elevados. Essa disparidade sistemática, **média muito maior que a mediana** sugere a existência de **observações extremas** que distorcem a média aritmética, fenômeno típico de distribuições com **outliers estruturais**.

Em outras palavras, o comportamento observado indica que:

- a maioria dos pedidos é processada de forma rápida (medianas baixas);
- porém, um pequeno conjunto de pedidos apresenta tempos **profundamente discrepantes**, produzindo **caudas longas** e elevando consideravelmente as médias;
- essa discrepância é consistente com a dinâmica real de processos logísticos, nos quais atrasos pontuais — embora raros — têm grande impacto no desempenho operacional e, consequentemente, nos indicadores de OTIF.

Assim, a assimetria à direita encontrada não é apenas estatisticamente esperada, mas também operacionalmente significativa, reforçando o entendimento de que **os atrasos logísticos são eventos esparsos, porém severos**, e constituem justamente o fenômeno que o modelo preditivo deverá capturar.

> ***Observações:*** foi destacado o tempo de processo apenas para facilitar a esposição, mas observa-se que existem a disparidade em todos as variaveis: no `peso` notodamos que até 75% dos dados pesam até 17.73 kg, ou seja 2/3 dos dados, contudo os 25% restantes se tramtam dos valores da ponta da cauda longa, que distorcem as médias com o valor máximo de 90.390 Kg.

# 3. Medidas de Dispersão (Variância, Desvio-Padrão e Coeficiente de Variação)

***Descrição:***
As medidas de dispersão avaliadas nesta etapa têm como objetivo quantificar **o grau de variabilidade** das variáveis numéricas. Enquanto as medidas de tendência central mostram *onde* os dados tendem a se concentrar, as medidas de dispersão mostram *o quanto* os dados se afastam desse centro.

Essas medidas são fundamentais para compreender:

- a estabilidade do processo operacional,
- a existência de variabilidade excessiva (ruído ou eventos críticos),
- o impacto de valores extremos nas médias,
- e o grau de risco estatístico associado a cada etapa do fluxo logístico.

As análises desta etapa complementam diretamente o estudo de assimetria e preparam o terreno para as próximas fases de correlação e inferência.

---

## **Medidas e Visualizações**

- **Variância (`var`)** — indica o quanto os valores se dispersam em relação à média.
- **Desvio-Padrão (`std`)** — raiz da variância; mostra a dispersão na mesma escala da variável.
- **Coeficiente de Variação (`CV = std / mean`)** — mede a variabilidade relativa, permitindo comparação entre variáveis de escalas diferentes.

---

### **Interpretações Importantes**

> - **Se std é alto** → grande instabilidade operacional; pode indicar gargalos, atrasos pontuais ou processos imprevisíveis.
> - **Se CV > 1.0** (100%) → dispersão maior que a média → forte volatilidade; comportamento caótico.
> - **Se CV entre 0.30 e 1.0** → variabilidade moderada; processos alternam entre normal e crítico.
> - **Se CV < 0.30** → variabilidade baixa; processo estável.

---

### **Sinalizações práticas para OTIF**

- Variáveis com **alto std** ou **alto CV** tendem a ser **ofensores operacionais**, contribuindo para atrasos.
- Variáveis com **baixo desvio** indicam processos mais previsíveis e estáveis.
- Comparar CV entre colunas poderá evidenciar quais etapas “puxam a cauda” da distribuição.